packages/node-sdk/paima-db/src/sql/extensions.sql

```diff
@@ -10,18 +10,21 @@ SELECT * FROM chain_data_extensions
 WHERE cde_name = :cde_name!;
 
 /* @name registerChainDataExtension */
-INSERT INTO chain_data_extensions(
-    cde_id,
-    cde_type,
-    cde_name,
-    cde_hash,
-    start_blockheight,
-    scheduled_prefix
-) VALUES (
-    :cde_id!,
+INSERT INTO
+    chain_data_extensions (
+        CDE_ID,
+        CDE_TYPE,
+        CDE_NAME,
+        CDE_HASH,
+        START_BLOCKHEIGHT,
+        SCHEDULED_PREFIX
+    )
+SELECT
+    COALESCE(:cde_id, MAX(chain_data_extensions.cde_id) + 1),
```

Comment:
I think there could actually be an issue with this, although I haven't tested this yet. Currently when loading the extensions on startup the engine will first load the ones from the file, and use the order to populate the `cdeId` field. Afterwards we load the dynamic extensions from the table. But if an extension gets added to `extensions.yml` I think the `cdeId` will clash with the one from the first dynamic primitive.

I'm not really sure how to handle this from the engine's point of view though, since it's more or less assumed that the `extensions.yml` shouldn't change.

I guess a way of making the Tarochi flow work could be to remove the dynamic primitives from the table, and adding them to `extensions.yml` before the new primitive.

Reply:
The Tarochi flow is hacky and shouldn't be a long-term solution. Games need to update occasionally and that may require occasionally adding new primitive so ideally we would have better support for that. Having to do stuff manually for Tarochi is okay since we know how things work, but obviously this wouldn't really be an option for external teams

Reply:
Yeah, I agree with that. It's just that currently there is no way of adding a primitive, so it's hard to say what that would look like.

Besides what I said before, there is also the fact that if you were allowed to edit the config file and the engine did a re-sync only for that primitive in order to get the state up-to-date, if you later did a re-sync from scratch then the `cde_id`s would also be different, since at that point we don't even know how many dynamic primitives would be there. This may not matter if nothing relies on this though, but it breaks determinism I guess (ignoring the fact that currently that would break determinism anyway because the new primitive may trigger state transitions in the past).

I guess something we could do here to future proof this is to assign cde ids to the dynamic primitives in a range that won't ever be used by the static ones? (negative numbers even?).

Reply:
> I guess something we could do here to future proof this is to assign cde ids to the dynamic primitives in a range that won't ever be used by the static ones? (negative numbers even?).

I think this will only partially be a solution because if you add a new dynamic primitive as part of an update, you may run into the same issue again, right?

Reply:
> I think this will only partially be a solution because if you add a new dynamic primitive as part of an update, you may run into the same issue again, right?

I don't follow why. What I'm saying is that the primitives in `extensions.yml` get assigned ids according to the order in the list like we do right now, for example in the range `[0; 100)`

And extensions created as a result of a dynamic trigger receive for example cde ids in the range `[100; ...)`. Of course using 100 just as an example to have small numbers.

Then if you start with a dynamic primitive trigger in your config, and 1 primitive is added dynamically, the ids `0` and `100` would be used. 

If you now upgrade and add a new trigger, it would have id `2`, so there is no collision there. And generated primitives would still be sequential, so the next one would be `101`.

Reply:
I think with the current code there should be a solution for this that only involves getting the cde ids from the cde name instead of relying on the order (except in the first-time setup). And if the engine knows that a primitive is new, then it's just a matter of querying the db to find the proper id.

But I think the name doesn't have to unique? And it doesn't solve the issue of a re-sync from scratch (if it's even an issue). It would also allow re-ordering the file, which may not be desired.

Reply:
> I don't follow why. 

Imagine your you have a dynamic primitive A in your codebase.
- At block 1111, A creates dynamic primitive ID 100
- At block 2111, A creates dynamic primitive ID 101

Now, you want to modify your game to include a new dynamic primitive B monitoring an existing contract for a new functionality. It's possible that this new dynamic primitive B has something at block 1112, which means you now end up with

- At block 1111, A creates dynamic primitive ID 100
- At block 1112, B creates *new* dynamic primitive ID 101
- At block 2111, A creates dynamic primitive ID **102** (note: this is no longer the same ID as before)

Reply:
I see, right, that's a good point. That said, I don't think we really care about this as long as the id's are consistent in the db. I did some searching and I couldn't find any place where the id's are exposed to the game. It's certainly a problem if a new primitive got assigned the id of an existing one, which is what would happen right now if we naively allowed adding a primitive to `extensions.yml`,  but if the same primitive has a different id it shouldn't really matter as long as there are no clashes I believe.

The only potential difference is that it may lead to a different order in the extension processing maybe? Although currently the dynamic primitives are not sorted by any field, I imagine the sorting may be that of the primary key. But not sure if anything relies on this order.

Reply:
I think one way we can solve this with a similar solution is to say there are at most 32-bits work of cde IDs in the extensions.yml. Then, we say dynamic primitives always start with 4 bits of the parent ID, following by 48 bits for the order in which it was added (48 because max safe ints in Javascript are 52 bits)

42bits = 4,398,046,511,103 which should be more than enough. This also guarantees there are no conflicts

Reply:
why 4 bits? doesn't that limit you to 16 extensions? The 0th and the 16th would have the same 4 bits, so they would get the same namespace. I think for this to make sense we would need to use 52 - 32 = 20 bits for the split. I feel like 2 ** 20 = 1,048,576 dynamic primitives per dynamic trigger it's already a lot anyway.

Btw, we could also extract the bits from the hash... But I guess at that point we could just stop using incrementing ids and just use the hash as the id.

Reply:
Oh yeah my bad. It would have to be something more like 16 bits

Instead of using an incrementing number, using the hash of a primitive's configuration as an ID and not allowing 2 primitives with the same hash also sounds like it makes sense